flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/MapRangePartitioner.java

```diff
@@ -0,0 +1,288 @@
+      mapStatistics.forEach(
+          (k, v) -> {
+            int estimatedSplits = (int) Math.ceil(v / targetWeightPerSubtask);
+            long estimatedCloseFileCost = closeFileCostInWeight * estimatedSplits;
+            sortedStatsWithCloseFileCost.put(k, v + estimatedCloseFileCost);
+          });
+
+      long totalWeightWithCloseFileCost =
+          sortedStatsWithCloseFileCost.values().stream().mapToLong(l -> l).sum();
+      long targetWeightPerSubtaskWithCloseFileCost =
+          (long) Math.ceil(((double) totalWeightWithCloseFileCost) / numPartitions);
+      this.assignment =
+          buildAssignment(
+              numPartitions, sortedStatsWithCloseFileCost, targetWeightPerSubtaskWithCloseFileCost);
+    }
+
+    return assignment;
+  }
+
+  @VisibleForTesting
+  NavigableMap<SortKey, Long> sortedStatsWithCloseFileCost() {
+    return sortedStatsWithCloseFileCost;
+  }
+
+  /**
+   * @return assignment summary for every subtask. Key is subtaskId. Value pair is (weight assigned
+   *     to the subtask, number of keys assigned to the subtask)
+   */
+  Map<Integer, Pair<Long, Integer>> assignmentInfo() {
+    Map<Integer, Pair<Long, Integer>> assignmentInfo = Maps.newTreeMap();
+    assignment.forEach(
+        (key, keyAssignment) -> {
+          for (int i = 0; i < keyAssignment.assignedSubtasks.length; ++i) {
+            int subtaskId = keyAssignment.assignedSubtasks[i];
+            long subtaskWeight = keyAssignment.subtaskWeights[i];
+            Pair<Long, Integer> oldValue = assignmentInfo.getOrDefault(subtaskId, Pair.of(0L, 0));
+            assignmentInfo.put(
+                subtaskId, Pair.of(oldValue.first() + subtaskWeight, oldValue.second() + 1));
+          }
+        });
+
+    return assignmentInfo;
+  }
+
+  private Map<SortKey, KeyAssignment> buildAssignment(
+      int numPartitions,
+      NavigableMap<SortKey, Long> sortedStatistics,
+      long targetWeightPerSubtask) {
+    Map<SortKey, KeyAssignment> assignmentMap =
+        Maps.newHashMapWithExpectedSize(sortedStatistics.size());
+    Iterator<SortKey> mapKeyIterator = sortedStatistics.keySet().iterator();
+    int subtaskId = 0;
+    SortKey currentKey = null;
+    long keyRemainingWeight = 0L;
+    long subtaskRemainingWeight = targetWeightPerSubtask;
+    List<Integer> assignedSubtasks = Lists.newArrayList();
+    List<Long> subtaskWeights = Lists.newArrayList();
+    while (mapKeyIterator.hasNext() && subtaskId < numPartitions) {
+      if (currentKey == null) {
+        currentKey = mapKeyIterator.next();
+        keyRemainingWeight = sortedStatistics.get(currentKey);
+      }
+
+      assignedSubtasks.add(subtaskId);
+      // assign the remaining weight of key to the current subtask if it is the last subtask
+      // or if the subtask has more capacity than the remaining key weight
+      if (subtaskId == numPartitions - 1 || keyRemainingWeight < subtaskRemainingWeight) {
```

Comment:
If I understand correctly `keyRemainingWeight < subtaskRemainingWeight` should always be true for  `subtaskId == numPartitions - 1`.
How confident is you in the algorithm above? (I did not find any issue, but...) Would it worth to log a message at least if something is off, and we put keys to the last task just because we made some issue during calculation?

Reply:
> If I understand correctly keyRemainingWeight < subtaskRemainingWeight should always be true for subtaskId == numPartitions - 1.

not sure I fully understand the comment. this is an `or` condition. fully assign the remaining key weight to the subtask

1. if it is the last subtask
2. (or) if the weight is less than the subtask remaining capacity

Reply:
If we done the calculation correctly, then even for the last subtask, the size of the last subtask should be smaller than the `subtaskRemainingWeight`
If we depend on the `subtaskId == numPartitions - 1` part of the `if` clause, then we have a `wrong` distribution.

Reply:
we compute the target weigh using `ceiling` function. so last subtask should only get less or equal to the fair share.
```
long targetWeightPerSubtaskWithCloseFileCost =
          (long) Math.ceil(((double) totalWeightWithCloseFileCost) / numPartitions);
```

I agree that we shouldn't need `subtaskId == numPartitions - 1` in theory.  it was added for extra safety.  Please let me know your opinion on the current behavior (option 1) vs the alternatives below.

Option 2: log an error while maintaining the permissive behavior
```
if (subtaskId == numPartitions - 1 && keyRemainingWeight > subtaskRemainingWeight) {
    LOG.error("Invalid assignment: last subtask is assigned more weight than target");
}
```

Option 3: throw an exception
```
if (subtaskId == numPartitions - 1 && keyRemainingWeight > subtaskRemainingWeight) {
    throw new InvalidStateException("Invalid assignment: last subtask is assigned more weight than target");
}

```

Reply:
I would go for the option 2, and maybe a metric?

With option 3 we just restart the job. The first checkpoint will run without a known distribution, and the job would probably continue to run without an issue.

Both option 2 and 3 requires conscious monitoring from the job owners, and option 2 is better in many ways

Reply:
thanks. will go with option 2 of logging an error msg then

Reply:
> the job is constantly restarting 

Do we store the distribution in the state? That’s the reason why the job will fail again after a restart?